source/server/connection_handler_impl.cc

```diff
@@ -95,6 +95,36 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
           details->typed_listener_)) {
     tcp_listener_map_by_address_.insert_or_assign(
         config.listenSocketFactory().localAddress()->asStringView(), details);
+
+    auto& address = details->address_;
+    // If the address is Ipv6 and isn't v6only, parse out the ipv4 compatible address from the Ipv6
+    // address and put an item to the map. Then this allows the `getBalancedHandlerByAddress`
+    // can match the Ipv4 request to Ipv4-mapped address also.
+    if (address->type() == Network::Address::Type::Ip &&
+        address->ip()->version() == Network::Address::IpVersion::v6 &&
+        !address->ip()->ipv6()->v6only()) {
+      if (address->ip()->isAnyAddress()) {
+        // Since both "::" with ipv4_compat and "0.0.0.0" can be supported.
+        // If there already one and it isn't shutdown for compatible addr,
+        // then won't insert a new one.
+        auto ipv4_any_address =
+            Network::Address::Ipv4Instance(address->ip()->port()).asStringView();
+        auto ipv4_any_listener = tcp_listener_map_by_address_.find(ipv4_any_address);
+        if (ipv4_any_listener == tcp_listener_map_by_address_.end() ||
+            ipv4_any_listener->second->listener_->listener() == nullptr) {
+          tcp_listener_map_by_address_.insert_or_assign(ipv4_any_address, details);
```

Comment:
Can you guard this with the runtime behavior? 

My concern is that the new matching behavior, though correct, will break a lot of existing [::] listener configs

because the new behavior requires **both** the `filter chains` ipv4 set and ipv6 set. I don't think all the control planes are that strong.

CC @howardjohn

Reply:
Emm...this is the regression https://github.com/envoyproxy/envoy/issues/18134 we are trying to fix. So...it should be a regression, not a behavior change.

Reply:
@soulxu  Make sense!